internal/widget/termgrid.go

```diff
@@ -0,0 +1,241 @@
+package widget
+
+import (
+	"image/color"
+
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+// TermGrid is a monospaced grid of characters.
+// This is designed to be used in a terminal emulator.
+type TermGrid struct {
+	widget.TextGrid
+}
+
+// NewTermGrid creates a new empty TermGrid widget.
+func NewTermGrid() *TermGrid {
+	tg := &TermGrid{}
+	tg.ExtendBaseWidget(tg)
+	return tg
+}
+
+// HighlightRange highlight options to the given range
+// if highlighting has previously been applied it is enabled
+func (t *TermGrid) HighlightRange(blockMode bool, startRow, startCol, endRow, endCol int, o ...HighlightOption) {
+	applyHighlight := func(cell *widget.TextGridCell) {
+		// Check if already highlighted
+		if h, ok := cell.Style.(*HighlightedTextGridStyle); !ok {
+			highlightedStyle := &HighlightedTextGridStyle{OriginalStyle: cell.Style, Highlighted: true}
+			highlightedStyle.With(o...)
+			cell.Style = highlightedStyle
+		} else {
+			h.Highlighted = true
+		}
+	}
+
+	t.forRange(blockMode, startRow, startCol, endRow, endCol, applyHighlight, nil)
+}
+
+// ClearHighlightRange disables the highlight style for the given range
+func (t *TermGrid) ClearHighlightRange(blockMode bool, startRow, startCol, endRow, endCol int) {
+	clearHighlight := func(cell *widget.TextGridCell) {
+		// Check if already highlighted
+		if h, ok := cell.Style.(*HighlightedTextGridStyle); ok {
+			h.Highlighted = false
+		}
+	}
+	t.forRange(blockMode, startRow, startCol, endRow, endCol, clearHighlight, nil)
+}
+
+// HighlightedTextGridStyle defines a style that can be original or highlighted.
+type HighlightedTextGridStyle struct {
+	OriginalStyle    widget.TextGridStyle
+	HighlightedStyle widget.TextGridStyle
+	Highlighted      bool
+}
+
+// TextColor returns the color of the text, depending on whether it is highlighted.
+func (h *HighlightedTextGridStyle) TextColor() color.Color {
+	if h.Highlighted {
+		return h.HighlightedStyle.TextColor()
+	}
+	return h.OriginalStyle.TextColor()
+}
+
+// BackgroundColor returns the background color, depending on whether it is highlighted.
+func (h *HighlightedTextGridStyle) BackgroundColor() color.Color {
+	if h.Highlighted {
+		return h.HighlightedStyle.BackgroundColor()
+	}
+	return h.OriginalStyle.BackgroundColor()
+}
+
+// HighlightOption defines a function type that can modify a HighlightedTextGridStyle.
+type HighlightOption func(h *HighlightedTextGridStyle)
+
+// InvertColor inverts a color c with the given bitmask
+func InvertColor(c color.Color, bitmask uint8) color.Color {
+	r, g, b, a := c.RGBA()
+	return color.RGBA{
+		R: uint8(r>>8) ^ bitmask,
+		G: uint8(g>>8) ^ bitmask,
+		B: uint8(b>>8) ^ bitmask,
+		A: uint8(a >> 8),
+	}
+}
+
+// WithInvert returns a HighlightOption that inverts the colors of the HighlightedTextGridStyle using the provided bitmask.
+func WithInvert(bitmask uint8) HighlightOption {
```

Comment:
In the context of `widget` package this function name does not seem complete.

That said I'm not sure the type needs to be exported at all - isn't the highlight always going to use the invert option? Feels like added API (and a whole new pattern) for no current requirement.

Reply:
I actually re-write most of this while implementing the "blink" and underline functionality. 
I could break some of that into here.

Reply:
FYI https://github.com/codelaboratoryltd/terminal/commit/c15ca200736cc054eabd5188884c9484c9500cd9